### ISSUE TITLE
ignoreExternalBonds should still build external bonds that are present

### ISSUE THREAD

**Issue by tristanic** [pandegroup/openmm#2096]
When I first suggested the addition of this flag, I called it 'ignoreMissingExternalBonds' with the idea that external bonds should still be applied where present, but residues that would be completed by an external bond that's _not_ present should be allowed to continue forward. That is indeed how it works for the AMBER forcefield, but if I use charmm36.xml disulfides are not properly created - in the resulting simulation the paired sulfur atoms are able to pass right through each other. I understand this is a somewhat niche application, but it would be good to work out a solution.

**Comment by peastman** [comment]
The only effect of `ignoreExternalBonds` is on matching templates to determine atom types.  Once that's done, it has no further effect.  It will still attempt to create a bond for every bond listed in your topology, regardless of whether that bond is internal or external to the residue.  The only requirement is that it finds parameters definitions for the atom types that got assigned.

I think this is probably what's happening.  There are two possible versions of CYS that each have identical atoms and internal bonds: one that takes part in a disulfide bond, and one that is positively charged.  These can be distinguished by the presence or absence of an external bond connected to the SG atom.  But you've told it to ignore external bonds, so it's arbitrary which of the two templates will get picked.  Those two templates assign different atom types to the SG atom.  If it picks the wrong one, there won't be any parameters defined for a bond between the two SG atoms, so no bond will get added.

The documentation for `ignoreExternalBonds` mentions this possibility and what to do about it:

> This option can create ambiguities where multiple templates match the same residue.  If that happens, use the residueTemplates argument to specify which one to use.

**Comment by tristanic** [comment]
Ah - OK, that’s what I’m doing already for AMBER (where forcefield.py throws an error if I don’t provide templates for each Cys). Is this going to work correctly for CHARMM, though? The difference here is that there *is* no pre-existing template for a disulfide-bonded Cys - the different templates are created on-the-fly in forcefield.py by applying the different possible patches, so there’s no template name to provide.

 
 
Tristan Croll
Research Fellow
Cambridge Institute for Medical Research
University of Cambridge CB2 0XY

**Comment by tristanic** [comment]
Yeah... have you gotten to the glycan patches yet? Those get really painful - you either have to rely on residue name, or explicitly take geometry into account (different patches for alpha- and beta-glycosidic bonds). I guess that’s also true for sugars in general, though: alpha- and beta-mannose have the same topology but different geometry, for example. Ouch.

 
 
Tristan Croll
Research Fellow
Cambridge Institute for Medical Research
University of Cambridge CB2 0XY

**Comment by jchodera** [comment]
> Yeah... have you gotten to the glycan patches yet? Those get really painful - you either have to rely on residue name, or explicitly take geometry into account (different patches for alpha- and beta-glycosidic bonds). I guess that’s also true for sugars in general, though: alpha- and beta-mannose have the same topology but different geometry, for example. Ouch.

We haven't tackled the AMBER glycam force field yet, but the CHARMM36 carbohydrates are being included in the forcefield conversion right now.

**Comment by tristanic** [comment]
> We haven't tackled the AMBER glycam force field yet, but the CHARMM36 carbohydrates are being included in the forcefield conversion right now.

Great to know! I'll really look forward to that.

If I remove the offending disulfides in 1pmx and run with `ignoreExternalBonds=False`, CHARMM36 fails on the N-terminal amine:


>ValueError: No template found for residue 1 (GLY). The set of atoms matches ACEM, but the bonds are different. Perhaps the chain is missing a terminal group?


A thought on how to deal with such issues: at present forcefield.py attempts to match against standard residues first, then creates the signatures for all possible patches and tries to match against those. What if instead the patched signatures were created up-front and given descriptive names such as RESNAME-PATCHNAME, and added to the general template database? That would both solve the above N-terminal problem, and provide accessible hooks for the `residueTemplates` argument.

**Comment by Astrovicis** [comment]
Early awaiting a resolution to this one 😉
